File: c++/QEasyWidgets/Common/QFunctions.cpp
#include "QFunctions.h"

#include <QGraphicsOpacityEffect>
#include <QEasingCurve>
#include <QDir>
#include <QDebug>
#include <QMainWindow>
#include <QDockWidget>


void removeSubWidgets(QStackedWidget *widget) {
    if (!widget) return;

    while (widget->count() > 0) {
        QWidget *w = widget->widget(0);
        widget->removeWidget(w);
        w->deleteLater();
    }
}


void setRetainSizeWhenHidden(QWidget *widget, bool retainSize) {
    if (!widget) return;

    QSizePolicy policy = widget->sizePolicy();
    policy.setRetainSizeWhenHidden(retainSize);
    widget->setSizePolicy(policy);
}


void setDropShadowEffect(QWidget *widget, qreal radius, const QColor &color) {
    if (!widget) return;

    QGraphicsDropShadowEffect *effect = new QGraphicsDropShadowEffect(widget);
    effect->setOffset(0, 0);
    effect->setBlurRadius(radius);
    effect->setColor(color);
    widget->setGraphicsEffect(effect);
}


void setAnimation(QPropertyAnimation *animation, const QVariant &startValue, const QVariant &endValue, int duration) {
    if (!animation) return;

    animation->setStartValue(startValue);
    animation->setEndValue(endValue);
    animation->setDuration(duration);
    animation->setEasingCurve(QEasingCurve::InOutQuart);
}


QPropertyAnimation* setWidgetPosAnimation(QWidget *widget, int duration) {
    if (!widget) return nullptr;

    QRect originalGeometry = widget->geometry();
    QRect alteredGeometry(
        originalGeometry.left(),
        originalGeometry.top() + originalGeometry.height() / duration,
        originalGeometry.width(),
        originalGeometry.height()
    );

    QPropertyAnimation *animation = new QPropertyAnimation(widget, "geometry", widget);
    setAnimation(animation, originalGeometry, alteredGeometry, duration);

    return animation;
}


QParallelAnimationGroup* setWidgetSizeAnimation(QWidget *widget, int targetWidth, int targetHeight, int duration) {
    if (!widget) return nullptr;

    int currentWidth = getWidth(widget);
    int currentHeight = getHeight(widget);

    QParallelAnimationGroup *animationGroup = new QParallelAnimationGroup(widget);

    if (widget->property("currentWidth").isValid() && widget->property("currentHeight").isValid()) {
        if (targetWidth >= 0) {
            QPropertyAnimation *widthAnim = new QPropertyAnimation(widget, "currentWidth", widget);
            setAnimation(widthAnim, currentWidth, targetWidth, duration);
            animationGroup->addAnimation(widthAnim);
        }
        if (targetHeight >= 0) {
            QPropertyAnimation *heightAnim = new QPropertyAnimation(widget, "currentHeight", widget);
            setAnimation(heightAnim, currentHeight, targetHeight, duration);
            animationGroup->addAnimation(heightAnim);
        }
    } else {
        if (targetWidth >= 0) {
            QPropertyAnimation *minWidthAnim = new QPropertyAnimation(widget, "minimumWidth", widget);
            QPropertyAnimation *maxWidthAnim = new QPropertyAnimation(widget, "maximumWidth", widget);
            setAnimation(minWidthAnim, currentWidth, targetWidth, duration);
            setAnimation(maxWidthAnim, currentWidth, targetWidth, duration);
            animationGroup->addAnimation(minWidthAnim);
            animationGroup->addAnimation(maxWidthAnim);
        }
        if (targetHeight >= 0) {
            QPropertyAnimation *minHeightAnim = new QPropertyAnimation(widget, "minimumHeight", widget);
            QPropertyAnimation *maxHeightAnim = new QPropertyAnimation(widget, "maximumHeight", widget);
            setAnimation(minHeightAnim, currentHeight, targetHeight, duration);
            setAnimation(maxHeightAnim, currentHeight, targetHeight, duration);
            animationGroup->addAnimation(minHeightAnim);
            animationGroup->addAnimation(maxHeightAnim);
        }
    }

    return animationGroup;
}


QPropertyAnimation* setWidgetOpacityAnimation(QWidget *widget, qreal originalOpacity, qreal targetOpacity, int duration) {
    if (!widget) return nullptr;

    QGraphicsOpacityEffect *effect = new QGraphicsOpacityEffect(widget);
    widget->setGraphicsEffect(effect);

    QPropertyAnimation *animation = new QPropertyAnimation(effect, "opacity", widget);
    setAnimation(animation, originalOpacity, targetOpacity, duration);

    return animation;
}


void setFont(QWidget *widget, int fontSize, QFont::Weight weight, const QStringList &families) {
    if (!widget) return;

    QFont font;
    font.setFamilies(families);
    font.setPixelSize(fontSize);
    font.setWeight(weight);
    widget->setFont(font);
}


void setText(QWidget *widget, const QString &text, bool setHtml, bool setPlaceholderText, const QString &placeholderText) {
    if (!widget) return;

    // Try different text setting methods
    if (widget->metaObject()->indexOfMethod("setText(QString)") >= 0) {
        QMetaObject::invokeMethod(widget, "setText", Q_ARG(QString, text));
    }
    if (widget->metaObject()->indexOfMethod("setPlainText(QString)") >= 0) {
        QMetaObject::invokeMethod(widget, "setPlainText", Q_ARG(QString, text));
    }
    if (setHtml && widget->metaObject()->indexOfMethod("setHtml(QString)") >= 0) {
        QMetaObject::invokeMethod(widget, "setHtml", Q_ARG(QString, text));
    }
    if (setPlaceholderText && widget->metaObject()->indexOfMethod("setPlaceholderText(QString)") >= 0) {
        QString phText = placeholderText.isEmpty() ? text : placeholderText;
        QMetaObject::invokeMethod(widget, "setPlaceholderText", Q_ARG(QString, phText));
    }
}


QString getText(QWidget *widget, bool getHtml, bool getPlaceholderText) {
    if (!widget) return QString();

    QString text;

    // Try different text getting methods
    if (widget->metaObject()->indexOfMethod("text()") >= 0) {
        QMetaObject::invokeMethod(widget, "text", Q_RETURN_ARG(QString, text));
    } else if (widget->metaObject()->indexOfMethod("toPlainText()") >= 0) {
        QMetaObject::invokeMethod(widget, "toPlainText", Q_RETURN_ARG(QString, text));
    }

    if (getHtml && widget->metaObject()->indexOfMethod("toHtml()") >= 0) {
        QMetaObject::invokeMethod(widget, "toHtml", Q_RETURN_ARG(QString, text));
    }
    if (getPlaceholderText && widget->metaObject()->indexOfMethod("placeholderText()") >= 0) {
        QString phText;
        QMetaObject::invokeMethod(widget, "placeholderText", Q_RETURN_ARG(QString, phText));
        if (!phText.isEmpty() && text.trimmed().isEmpty()) {
            text = phText;
        }
    }

    return text;
}


QString getFileDialog(FileDialogMode mode, const QString &fileType, const QString &directory) {
    QString result;
    QString startDir = directory.isEmpty() ? QDir::currentPath() : directory;

    if (!QDir(startDir).exists()) {
        QDir().mkpath(startDir);
    }

    switch (mode) {
        case FileDialogMode::SelectFolder:
            result = QFileDialog::getExistingDirectory(nullptr, "选择文件夹 | SelectFolder", startDir);
            break;
        case FileDialogMode::SelectFile: {
            QString filter = fileType.isEmpty() ? "任意类型 (*.*)" : fileType;
            result = QFileDialog::getOpenFileName(nullptr, "选择文件 | SelectFile", startDir, filter);
            break;
        }
        case FileDialogMode::SaveFile: {
            QString filter = fileType.isEmpty() ? "任意类型 (*.*)" : fileType;
            result = QFileDialog::getSaveFileName(nullptr, "保存文件 | SaveFile", startDir, filter);
            break;
        }
    }

    return result;
}


void setContextMenu(QWidget *parent, QMenu *contextMenu, const QMap<QString, std::function<void()>> &actions) {
    if (!parent || !contextMenu) return;

    for (auto it = actions.begin(); it != actions.end(); ++it) {
        QAction *action = new QAction(it.key(), parent);
        // Connect using lambda to call the std::function
        QObject::connect(action, &QAction::triggered, parent, it.value());
        contextMenu->addAction(action);
    }
}


void showContextMenu(QWidget *parent, QMenu *contextMenu, const QMap<QString, std::function<void()>> &actions, const QPoint &position) {
    setContextMenu(parent, contextMenu, actions);
    contextMenu->exec(position);
}


void openURL(const QString &url, bool createIfNotExist) {
    QUrl qurl = QUrl::fromLocalFile(url);
    if (qurl.isValid()) {
        if (createIfNotExist) {
            QDir().mkpath(url);
        }
        bool success = QDesktopServices::openUrl(qurl);
        if (!success) {
            qDebug() << "Failed to open URL:" << url;
        }
    } else {
        qDebug() << "Invalid URL:" << url;
    }
}


void openURL(const QStringList &urls, bool createIfNotExist) {
    for (const QString &url : urls) {
        openURL(url, createIfNotExist);
    }
}


void saveLayout(QWidget *widget, QSettings *settings) {
    if (!widget || !settings) return;

    settings->setValue("layout/geometry", widget->saveGeometry());
    if (QMainWindow *mainWindow = qobject_cast<QMainWindow*>(widget)) {
        settings->setValue("layout/state", mainWindow->saveState());
    }
    settings->sync();
}


void restoreLayout(QWidget *widget, QSettings *settings) {
    if (!widget || !settings) return;

    widget->restoreGeometry(settings->value("layout/geometry").toByteArray());
    if (QMainWindow *mainWindow = qobject_cast<QMainWindow*>(widget)) {
        mainWindow->restoreState(settings->value("layout/state").toByteArray());
        // Restore dock widgets
        QList<QDockWidget*> dockWidgets = mainWindow->findChildren<QDockWidget*>();
        for (QDockWidget *dock : dockWidgets) {
            mainWindow->restoreDockWidget(dock);
        }
    }
}


QScreen* getCurrentScreen() {
    QPoint cursorPos = QCursor::pos();
    QList<QScreen*> screens = QApplication::screens();
    for (QScreen *screen : screens) {
        if (screen->geometry().contains(cursorPos)) {
            return screen;
        }
    }
    return QApplication::primaryScreen();
}


QRect getScreenGeometry(QScreen *screen, bool getAvailableGeometry) {
    if (!screen) {
        screen = getCurrentScreen();
        if (!screen) {
            screen = QApplication::primaryScreen();
        }
    }

    if (!screen) return QRect();

    return getAvailableGeometry ? screen->availableGeometry() : screen->geometry();
}

File: python/QEasyWidgets/Common/QFunctions.py
import os
from pathlib import Path
from typing import Union, Optional, Sequence
from PyEasyUtils import toIterable, normPath, runCMD
from PySide6.QtCore import Qt, QSettings, QPoint, QRect, QSize, QPropertyAnimation, QParallelAnimationGroup, QEasingCurve, QUrl
from PySide6.QtGui import QColor, QRgba64, QFont, QScreen, QDesktopServices, QAction, QCursor
from PySide6.QtWidgets import *

##############################################################################################################################

def findChild(
    parent: QWidget,
    childType: object
):
    """
    Function to find child widget
    """
    parentWidget_Children = parent.children()

    for parentWidget_Child in parentWidget_Children:
        if isinstance(parentWidget_Child, childType):
            return parentWidget_Child


def findParent(
    child: QWidget,
    parentType: object
):
    """
    Function to find parent widget
    """
    childWidget_Parent = child.parent()

    while not isinstance(childWidget_Parent, parentType):
        try:
            childWidget_Parent = childWidget_Parent.parent()
        except:
            childWidget_Parent = None
            break

    return childWidget_Parent

##############################################################################################################################

def getWidth(widget: QWidget):
    return widget.geometry().width() if widget.size() == QSize(100, 30) else widget.width()


def getHeight(widget: QWidget):
    return widget.geometry().height() if widget.size() == QSize(100, 30) else widget.height()

##############################################################################################################################

def removeSubWidgets(
    widget: QWidget
):
    if isinstance(widget, QStackedWidget):
        while widget.count():
            widget.removeWidget(widget.widget(0))


def setRetainSizeWhenHidden(
    widget: QWidget,
    retainSize: bool = True
):
    sizePolicy = widget.sizePolicy()
    sizePolicy.setRetainSizeWhenHidden(retainSize)
    widget.setSizePolicy(sizePolicy)


def setDropShadowEffect(
    widget: QWidget,
    blurRadius: float = 3.,
    color: Union[QColor, QRgba64] = Qt.gray,
    offset_dx: float = 0.,
    offset_dy: float = 0.,
):
    dropShadowEffect = QGraphicsDropShadowEffect()
    dropShadowEffect.setOffset(0, 0)
    dropShadowEffect.setBlurRadius(blurRadius)
    dropShadowEffect.setColor(color)
    dropShadowEffect.setOffset(offset_dx, offset_dy)
    widget.setGraphicsEffect(dropShadowEffect)
    return dropShadowEffect


def setOpacityEffect(
    widget: QWidget,
    parent: Optional[QWidget] = None,
    duration: int = 123,
):
    opacityAnim = QPropertyAnimation(widget, b'windowOpacity', parent)
    opacityAnim.setDuration(duration)
    return opacityAnim

##############################################################################################################################

def setAnimation(
    animation: QPropertyAnimation,
    startValue,
    endValue,
    duration: int
):
    animation.setStartValue(startValue)
    animation.setEndValue(endValue)
    animation.setDuration(duration)
    animation.setEasingCurve(QEasingCurve.InOutQuart)
    return animation


def setWidgetPosAnimation(
    widget: QWidget,
    duration: int = 99
):
    OriginalGeometry = widget.geometry()
    AlteredGeometry = QRect(OriginalGeometry.left(), OriginalGeometry.top() + OriginalGeometry.height() / duration, OriginalGeometry.width(), OriginalGeometry.height())

    WidgetAnimation = QPropertyAnimation(widget, b"geometry", widget)

    return setAnimation(WidgetAnimation, OriginalGeometry, AlteredGeometry, duration)


def setWidgetSizeAnimation(
    frame: QWidget,
    targetWidth: Optional[int] = None,
    targetHeight: Optional[int] = None,
    duration: int = 210,
):
    """
    Function to animate widget size
    """
    currentWidth, currentHeight = getWidth(frame), getHeight(frame)

    animationGroup = QParallelAnimationGroup(frame)

    if frame.property("currentWidth") and frame.property("currentHeight"):
        frameAnimationWidth = QPropertyAnimation(frame, b"currentWidth", frame)
        frameAnimationHeight = QPropertyAnimation(frame, b"currentHeight", frame)
        animationGroup.addAnimation(setAnimation(frameAnimationWidth, currentWidth, targetWidth, duration)) if targetWidth is not None else None
        animationGroup.addAnimation(setAnimation(frameAnimationHeight, currentHeight, targetHeight, duration)) if targetHeight is not None else None
    else:
        frameAnimationMinWidth = QPropertyAnimation(frame, b"minimumWidth", frame)
        frameAnimationMaxWidth = QPropertyAnimation(frame, b"maximumWidth", frame)
        frameAnimationMinHeight = QPropertyAnimation(frame, b"minimumHeight", frame)
        frameAnimationMaxHeight = QPropertyAnimation(frame, b"maximumHeight", frame)
        animationGroup.addAnimation(setAnimation(frameAnimationMinWidth, currentWidth, targetWidth, duration)) if targetWidth is not None else None
        animationGroup.addAnimation(setAnimation(frameAnimationMaxWidth, currentWidth, targetWidth, duration)) if targetWidth is not None else None
        animationGroup.addAnimation(setAnimation(frameAnimationMinHeight, currentHeight, targetHeight, duration)) if targetHeight is not None else None
        animationGroup.addAnimation(setAnimation(frameAnimationMaxHeight, currentHeight, targetHeight, duration)) if targetHeight is not None else None

    return animationGroup


def setWidgetOpacityAnimation(
    widget: QWidget,
    originalOpacity: float,
    targetOpacity: float,
    duration: int = 99
):
    OpacityEffect = QGraphicsOpacityEffect()
    widget.setGraphicsEffect(OpacityEffect)

    originalOpacity = originalOpacity
    AlteredOpacity = targetOpacity

    WidgetAnimation = QPropertyAnimation(OpacityEffect, b"opacity", widget)

    return setAnimation(WidgetAnimation, originalOpacity, AlteredOpacity, duration)

##############################################################################################################################

def setFont(
    widget: QWidget,
    fontSize:int = 12,
    weight: QFont.Weight = QFont.Normal,
    families: Sequence[str] = ['Microsoft YaHei']
):
    """
    Set the font of widget
    """
    font = QFont()
    font.setFamilies(families)
    font.setPixelSize(fontSize)
    font.setWeight(weight)
    widget.setFont(font)


def setText(
    widget: QWidget,
    text: str,
    setHtml: bool = True,
    setPlaceholderText: bool = False,
    placeholderText: Optional[str] = None
):
    if hasattr(widget, 'setText'):
        widget.setText(text)
    if hasattr(widget, 'setPlainText'):
        widget.setPlainText(text)
    if hasattr(widget, 'setHtml') and setHtml:
        widget.setHtml(text)
    if hasattr(widget, 'setPlaceholderText') and setPlaceholderText:
        widget.setPlaceholderText(str(placeholderText) if text.strip() in ('', str(None)) else text)


def getText(
    widget: QWidget,
    getHtml: bool = False,
    getPlaceholderText: bool = False
):
    if hasattr(widget, 'text'):
        text = widget.text()
    if hasattr(widget, 'toPlainText'):
        text = widget.toPlainText()
    if hasattr(widget, 'toHtml') and getHtml:
        text = widget.toHtml()
    if hasattr(widget, 'placeholderText') and getPlaceholderText:
        text = widget.placeholderText() if text.strip() in ('', str(None)) else text
    return text

##############################################################################################################################

class FileDialogMode: 
    SelectFolder = 0
    SelectFile = 1
    SaveFile = 2


def getFileDialog(
    mode: FileDialogMode,
    fileType: Optional[str] = None,
    directory: Optional[str] = None
):
    displayText = ""
    os.makedirs(directory, exist_ok = True) if directory is not None and Path(directory).exists() == False else None
    if mode == FileDialogMode.SelectFolder:
        displayText = QFileDialog.getExistingDirectory(
            caption = "选择文件夹 | SelectFolder",
            dir = directory if directory is not None else os.getcwd()
        )
    if mode == FileDialogMode.SelectFile:
        displayText, _ = QFileDialog.getOpenFileName(
            caption = "选择文件 | SelectFile",
            dir = directory if directory is not None else os.getcwd(),
            filter = fileType if fileType is not None else '任意类型 (*.*)'
        )
    if mode == FileDialogMode.SaveFile:
        displayText, _ = QFileDialog.getSaveFileName(
            caption = "保存文件 | SaveFile",
            dir = directory if directory is not None else os.getcwd(),
            filter = fileType if fileType is not None else '任意类型 (*.*)'
        )
    return displayText# if displayText != '' else None

##############################################################################################################################

def setContextMenu(parent: QWidget, contextMenu: QMenu, actions: dict):
    for actionName, events in actions.items():
        action = QAction(actionName, parent)
        for event in toIterable(events):
            action.triggered.connect(event)
        contextMenu.addAction(action)
    return contextMenu


def showContextMenu(parent: QWidget, contextMenu: QMenu, actions: dict, position: QPoint):
    contextMenu = setContextMenu(parent, contextMenu, actions)
    contextMenu.exec(position)

##############################################################################################################################

def openURL(
    url: Union[str, list],
    createIfNotExist: bool = False
):
    """
    Function to open web/local url
    """
    def OpenURL(url):
        QURL = QUrl().fromLocalFile(normPath(url))
        if QURL.isValid():
            os.makedirs(normPath(url), exist_ok = True) if createIfNotExist else None
            isSucceeded = QDesktopServices.openUrl(QURL)
            runCMD([f'start {url}']) if not isSucceeded else None
        else:
            print(f"Invalid url: {url} !")

    if isinstance(url, str):
        OpenURL(url)
    else:
        URLList = toIterable(url)
        for Index, url in enumerate(URLList):
            #url = Function_ParamsChecker(URLList)[Index] if isinstance(url, QObject) else url
            OpenURL(url)

##############################################################################################################################

def saveLayout(widget: QWidget, settings: QSettings):
    settings.setValue("layout/geometry", widget.saveGeometry())
    if isinstance(widget, QMainWindow):
        settings.setValue("layout/state", widget.saveState())
    settings.sync()


def resetLayout(widget: QWidget, settings: QSettings):
    widget.restoreGeometry(settings.value("layout/geometry"))
    if isinstance(widget, QMainWindow):
        widget.restoreState(settings.value("layout/state"))
        for dockWidget in widget.findChildren(QDockWidget):
            widget.restoreDockWidget(dockWidget)

##############################################################################################################################

def getCurrentScreen():
    """
    Get current screen
    """
    cursorPos = QCursor.pos()
    for screen in QApplication.screens():
        if screen.geometry().contains(cursorPos):
            return screen


def getScreenGeometry(
    screen: Optional[QScreen] = None,
    getAvaliableGeometry: bool = True
):
    """
    Get screen geometry
    """
    if screen is None:
        screen = getCurrentScreen() or QApplication.primaryScreen()
    return screen.availableGeometry() if getAvaliableGeometry else screen.geometry()

##############################################################################################################################